dashboard/design/survey_summary_design.md

```diff
@@ -0,0 +1,103 @@
+
+**Cons:** Expensive. (This could be mitigated by building the pipeline from the end first as discussed below.)
+
+**Cost Estimate:** 3-4 weeks for the complete pipeline. 7 ± 2 days to implement step 3 (Summarize) and the first component of step 4 (Present) first, enough to support CSF 2019 surveys.
+
+### Detail design
+The pipeline has 4 main actions
+1. Ingest: Download data from JotForm server to CDO database.
+2. Query: Find survey data that matches lookup conditions.
+3. Summarize: Aggregate survey data.
+4. Present: Display survey summaries in dashboard.
+
+![Diagram](survey_summary_diagram.png)
+
+Some patterns that this design follow
+- _Open-Closed Principle_\
+  The pipeline is open for extension such as adding new components or adding different implementation of a component. However, it avoids, as best as it could, modifying existing code which could introduce regression. Following Single Responsibility principle will help with this.
+- _Single Responsibility Principle_\
+  Each method, class, module should have only 1 clear purpose. In a stronger statement, each should have only one reason to change.
+- _Dependency Inversion Principle_\
+  High-level modules (complex logic) should not depend on low-level modules (primary operations). Both should depend on abstractions. This allows high-level modules to easily switch to different low-level modules implementation.
+- _Template Method Pattern_\
+  Define the skeleton of an algorithm, deferring some steps. E.g. Query module (step 2) could use different implementations of Parser, Indexer and Filter depends on how data is stored in database. The specific implementations of those child components will be send to Query component as input parameters (dependency injection). Strategy pattern enable switching among implementations of the same component.
+- _Strategy Pattern_\
+  Define an interface common to a family of algorithms, e.g. survey result reducers, and make them interchangeable. This lets the algorithm vary independently from clients that use it.
+
+
+### Building from the end
+We don't have to start from step 1 (Ingest), we can start from step 3 (Summarize) and create adapters to connect it to the current pipeline.
+
+Specifically for CSF survey work, we will create an input adapter to retrieve survey data from the current data model, and an output adapter to convert summary result to format that the current UI view understands. Looking at it from a different angle, these adapters are just concrete implementations of Retriever and Decorator.
+
+
+# Recommendation
+Option 3. Build a new generic survey pipeline, starting from step 3(Summarize module) first and create adapters to plug it in the current pipeline. This approach allows us to take advantage of existing work and support 2019 CSF survey first. Then, gradually build out the rest of the pipeline later.
```

Comment:
I agree with this recommendation - this is necessary work, but we probably can't do it all at once, so building a small part at a time seems good.  Can we get any more specific on what we'd build first? Step 3 is fairly large.